### ISSUE TITLE
FileUriExposedException on Android 7

### ISSUE THREAD

**Issue by gotev** [spacecowboy/NoNonsense-FilePicker#115]
After selecting a file on Android 7, the following exception gets raised:
```java
android.os.FileUriExposedException: file:///storage/emulated/0/Download/12716320_1029986900381445_2238459154914449891_o.jpg exposed beyond app through ClipData.Item.getUri()
   at android.os.StrictMode.onFileUriExposed(StrictMode.java:1799)
   at android.net.Uri.checkFileUriExposed(Uri.java:2346)
   at android.content.ClipData.prepareToLeaveProcess(ClipData.java:832)
   at android.content.Intent.prepareToLeaveProcess(Intent.java:8909)
   at android.content.Intent.prepareToLeaveProcess(Intent.java:8894)
   at android.app.Activity.finish(Activity.java:5188)
   at android.app.Activity.finish(Activity.java:5208)
   at com.nononsenseapps.filepicker.AbstractFilePickerActivity.onFilesPicked(AbstractFilePickerActivity.java:165)
   at com.nononsenseapps.filepicker.AbstractFilePickerFragment.onClickOk(AbstractFilePickerFragment.java:285)
   at com.nononsenseapps.filepicker.AbstractFilePickerFragment$2.onClick(AbstractFilePickerFragment.java:192)
   at android.view.View.performClick(View.java:5609)
   at android.view.View$PerformClick.run(View.java:22262)
   at android.os.Handler.handleCallback(Handler.java:751)
   at android.os.Handler.dispatchMessage(Handler.java:95)
   at android.os.Looper.loop(Looper.java:154)
   at android.app.ActivityThread.main(ActivityThread.java:6077)
   at java.lang.reflect.Method.invoke(Native Method)
   at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```
From [FileUriExposedException](https://developer.android.com/reference/android/os/FileUriExposedException.html):

The exception that is thrown when an application exposes a file:// Uri to another app.

This exposure is discouraged since the receiving app may not have access to the shared path. For example, the receiving app may not have requested the READ_EXTERNAL_STORAGE runtime permission, or the platform may be sharing the Uri across user profile boundaries.

Instead, apps should use content:// Uris so the platform can extend temporary permission for the receiving app to access the resource.

This is only thrown for applications targeting N or higher. Applications targeting earlier SDK versions are allowed to share file:// Uri, but it's strongly discouraged.

**Comment by gotev** [comment]
[This post](http://stackoverflow.com/a/38858040) can be useful to solve this issue

**Comment by spacecowboy** [comment]
This has been reported previously in #107 and I failed to reproduce it. The library should be considered a part of your app, so the result shouldn't be passing your app boundary.

Are you doing anything different from the sample app?

**Comment by gotev** [comment]
I'm using the file picker in my [android upload service demo project](https://github.com/gotev/android-upload-service). To reproduce the issue you can check out the project and run the example app. The exception is raised when you press on the browse button. This is due to the fact that the file picker exposes a file URI instead of a Content URI on Android 24.

**Comment by spacecowboy** [comment]
thanks, I'll investigate

**Comment by spacecowboy** [comment]
Ah! Finally managed to reproduce it! The error only occurs in conjunction with `ALLOW_MULTIPLE`

E.g., this turns out to be fine:

```java
        Intent i = new Intent();
        i.setData(file);
        setResult(Activity.RESULT_OK, i);
        finish();
```

But not this

```java
        Intent i = new Intent();
        i.putExtra(EXTRA_ALLOW_MULTIPLE, true);

            ClipData clip = null;
            for (Uri file : files) {
                if (clip == null) {
                    clip = new ClipData("Paths", new String[]{},
                            new ClipData.Item(file));
                } else {
                    clip.addItem(new ClipData.Item(file));
                }
            }
            i.setClipData(clip);

        setResult(Activity.RESULT_OK, i);
        finish();
```

I'll prioritize a fix

**Comment by spacecowboy** [comment]
In fact, this seems to be a bug in android itself: https://code.google.com/p/android/issues/detail?can=2&start=0&num=100&q=&colspec=ID%20Status%20Priority%20Owner%20Summary%20Stars%20Reporter%20Opened&groupby=&sort=&id=220555

**Comment by gotev** [comment]
I've read the thread and I think this is intended to be like this for security reasons. [Here](https://inthecheesefactory.com/blog/how-to-share-access-to-file-with-fileprovider-on-android-nougat/en) there is a good explanation. The Android bug in my opinion is that the `FileUriExposedException` is not raised when you call `setData` in the Intent, but it's called only when you set clip data, so from what I've read, `FileProvider` should be always used when passing URIs outside the library. I can help you address this issue if you need some help.

**Comment by spacecowboy** [comment]
No need, I have a fix incoming.

But I don't agree about you reasoning. The current behavior should not raise a bug because the intent is never being passed to another app. Since this piece of code is a library, it should be considered part of the same permission space.

Note for instance that the library does not declare any read/write permissions. You have to do that in your outer app manifest. So if the library code is granted the read permission by your app, then it stands to reason that should be allowed to pass a Uri which depends on that permission back to the app. Anything else is madness.

Next version this is fixed by never returning URIs for multiple selections. Instead pure strings will be returned.

**Comment by gotev** [comment]
> But I don't agree about you reasoning. The current behavior should not raise a bug because the intent is never being passed to another app. Since this piece of code is a library, it should be considered part of the same permission space.

> Note for instance that the library does not declare any read/write permissions. You have to do that in your outer app manifest. So if the library code is granted the read permission by your app, then it stands to reason that should be allowed to pass a Uri which depends on that permission back to the app. Anything else is madness.

The permission is declared in the manifest of the app which uses the FilePicker, that's true, but if the permission was not given to the app at runtime before starting the file picker, it gets asked [here](https://github.com/spacecowboy/NoNonsense-FilePicker/blob/master/library/src/main/java/com/nononsenseapps/filepicker/FilePickerFragment.java#L77). The calling package for the permission in this case is [com.nononsenseapps.filepicker](https://github.com/spacecowboy/NoNonsense-FilePicker/blob/master/library/src/main/java/com/nononsenseapps/filepicker/FilePickerFragment.java#L7), and what happens is that an intent gets passed from one package to another, and in the case of file URI on Android 24, this raised the `FileUriExposedException`. It's exactly what happened in my demo app, in which I don't request the runtime permission before calling the FilePicker

**Comment by spacecowboy** [comment]
You may be right. In case you are, I reverted my changes and pulled the new version.

Because if you are right, at some point they'll fix the bug that `intent.setData` doesn't throw the exception... So it's best to guard for that in the fix as well.

**Comment by spacecowboy** [comment]
@gotev I have now made another version, and released it as `4.0.0-beta1` so I don't have to pull it again in case I find something terribly wrong with the approach.

I have tested it against your example app and it seems to work but I'd appreciate if you could verify it as well. 

Diff of changes in your app:

```diff
diff --git i/examples/app/app/build.gradle w/examples/app/app/build.gradle
index f536733..71264d1 100644
--- i/examples/app/app/build.gradle
+++ w/examples/app/app/build.gradle
@@ -27,7 +27,7 @@ dependencies {
     compile 'com.android.support:appcompat-v7:24.2.1'
     compile "com.jakewharton:butterknife:${butterKnifeVersion}"
     apt "com.jakewharton:butterknife-compiler:${butterKnifeVersion}"
-    compile 'com.nononsenseapps:filepicker:3.1.0'
+    compile 'com.nononsenseapps:filepicker:4.0.0-beta1'
     //compile 'net.gotev:uploadservice:3.0.3'
     //compile 'net.gotev:uploadservice-okhttp:3.0.3'
     //compile 'net.gotev:uploadservice-ftp:3.0.3'
diff --git i/examples/app/app/src/main/AndroidManifest.xml w/examples/app/app/src/main/AndroidManifest.xml
index 6bc8231..c1e0b0d 100644
--- i/examples/app/app/src/main/AndroidManifest.xml
+++ w/examples/app/app/src/main/AndroidManifest.xml
@@ -23,6 +23,16 @@
             </intent-filter>
         </activity>
 
+        <provider
+            android:name="android.support.v4.content.FileProvider"
+            android:authorities="${applicationId}.provider"
+            android:exported="false"
+            android:grantUriPermissions="true">
+            <meta-data
+                android:name="android.support.FILE_PROVIDER_PATHS"
+                android:resource="@xml/nnf_provider_paths"/>
+        </provider>
+
         <activity
             android:name=".BackHandlingFilePickerActivity"
             android:label="@string/select_file"
diff --git i/examples/app/app/src/main/java/net/gotev/uploadservicedemo/MainActivity.java w/examples/app/app/src/main/java/net/gotev/uploadservicedemo/MainActivity.java
index 4145f77..bd58973 100644
--- i/examples/app/app/src/main/java/net/gotev/uploadservicedemo/MainActivity.java
+++ w/examples/app/app/src/main/java/net/gotev/uploadservicedemo/MainActivity.java
@@ -1,10 +1,8 @@
 package net.gotev.uploadservicedemo;
 
 import android.app.Activity;
-import android.content.ClipData;
 import android.content.Intent;
 import android.net.Uri;
-import android.os.Build;
 import android.os.Bundle;
 import android.os.Environment;
 import android.support.v7.app.AppCompatActivity;
@@ -257,28 +255,18 @@ public class MainActivity extends AppCompatActivity implements UploadStatusDeleg
             List<Uri> resultUris = new ArrayList<>();
 
             if (data.getBooleanExtra(FilePickerActivity.EXTRA_ALLOW_MULTIPLE, false)) {
-                // For JellyBean and above
-                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
-                    ClipData clip = data.getClipData();
-
-                    if (clip != null) {
-                        for (int i = 0; i < clip.getItemCount(); i++) {
-                            resultUris.add(clip.getItemAt(i).getUri());
-                        }
-                    }
-
-                // For Ice Cream Sandwich
-                } else {
                     ArrayList<String> paths = data.getStringArrayListExtra(FilePickerActivity.EXTRA_PATHS);
 
                     if (paths != null) {
                         for (String path: paths) {
-                            resultUris.add(Uri.parse(path));
+                            Uri uri = Uri.parse(path);
+                            File file = com.nononsenseapps.filepicker.Utils.getFileForUri(uri);
+                            resultUris.add(Uri.fromFile(file));
                         }
                     }
-                }
             } else {
-                resultUris.add(data.getData());
+                File file = com.nononsenseapps.filepicker.Utils.getFileForUri(data.getData());
+                resultUris.add(Uri.fromFile(file));
             }
 
             StringBuilder absolutePathsConcat = new StringBuilder();
```

**Comment by gotev** [comment]
I can confirm it's working! Great job, thank you!
